### PR TITLE
hparams as dict [blocked by 1041]

### LINKED ISSUE #977: Support storing hparams as a dict
## 🚀 Feature
Right now, we assume `model.hparams` is an `argparse.Namespace`. We've had a number of requests to support `hparams` as a simple `dict`. Let's do it.

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?
Fixes #977 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/neptune.py [hunk 1]
```diff
@@ -164,8 +164,9 @@ def experiment(self) -> Experiment:
         return self._experiment
 
     @rank_zero_only
-    def log_hyperparams(self, params: argparse.Namespace):
-        for key, val in vars(params).items():
+    def log_hyperparams(self, params: Union[dict, Namespace]):
```
Comment (yukw777) [inline pytorch_lightning/loggers/neptune.py]: should it be `Union[Dict, Namespace]`?

File: pytorch_lightning/loggers/tensorboard.py [hunk 2]
```diff
@@ -100,14 +99,8 @@ def experiment(self) -> SummaryWriter:
...
     @rank_zero_only
-    def log_hyperparams(self, params: argparse.Namespace):
-        if params is None:
-            return
-
-        # in case converting from namespace
-        if isinstance(params, Namespace):
-            params = vars(params)
-        params = dict(params)
+    def log_hyperparams(self, params: Union[dict, Namespace]):
```
Comment (yukw777) [inline pytorch_lightning/loggers/tensorboard.py]: same here, should it be `Union[Dict, Namespace]`?

File: pytorch_lightning/loggers/test_tube.py [hunk 3]
```diff
@@ -92,10 +92,11 @@ def experiment(self) -> Experiment:
         return self._experiment
 
     @rank_zero_only
-    def log_hyperparams(self, params: argparse.Namespace):
+    def log_hyperparams(self, params: Union[dict, Namespace]):
```
Comment (yukw777) [inline pytorch_lightning/loggers/test_tube.py]: `Union[Dict, Namespace]`?

File: pytorch_lightning/loggers/wandb.py [hunk 4]
```diff
@@ -91,7 +91,8 @@ def watch(self, model: nn.Module, log: str = 'gradients', log_freq: int = 100):
         wandb.watch(model, log=log, log_freq=log_freq)
 
     @rank_zero_only
-    def log_hyperparams(self, params: argparse.Namespace):
+    def log_hyperparams(self, params: Union[dict, Namespace]):
```
Comment (yukw777) [inline pytorch_lightning/loggers/wandb.py]: `Union[Dict, Namespace]`?

File: tests/models/base.py [hunk 5]
```diff
@@ -42,14 +43,16 @@ class TestModelBase(LightningModule):
     interface
     """
 
-    def __init__(self, hparams, force_remove_distributed_sampler=False):
-        """
-        Pass in parsed HyperOptArgumentParser to the model
+    def __init__(self, hparams: Union[dict, Namespace], force_remove_distributed_sampler: bool = False):
```
Comment (yukw777) [inline tests/models/base.py]: `Union[Dict, Namespace]`?

**Comment by Borda** [comment]
> This is awesome! Maybe we shouldn't have fixed the doc around this haha.

well, unless we define the `dict` type like `Dict[str, Any]` it does not matter :]

File: pytorch_lightning/loggers/base.py [hunk 6]
```diff
@@ -41,6 +42,13 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None):
         """
         pass
 
+    def _convert_params(self, params: Union[dict, Namespace]) -> dict:
+        assert isinstance(params, (dict, Namespace))
```
Comment (neggert) [inline pytorch_lightning/loggers/base.py]: I'm not sure this is a good idea. It goes against the whole idea of duck typing. Won't this throw an exception if I try to use e.g. an OrderedDict or defaultdict?

File: tests/models/base.py [hunk 7]
```diff
@@ -42,14 +43,16 @@ class TestModelBase(LightningModule):
...
+        """Pass in parsed HyperOptArgumentParser to the model
         :param hparams:
         """
         # init superclass
         super().__init__()
-        self.hparams = hparams
+
+        # accept also dict and convert it to Namespace locally
+        assert isinstance(hparams, (dict, Namespace))
```
Comment (neggert) [inline tests/models/base.py]: Same comment here

File: pytorch_lightning/loggers/base.py (hunk 6 above)
Comment (Borda) [reply pytorch_lightning/loggers/base.py]: Not sure, but typing is also not strict, right?
Ok I will remove it and extend Dict

**Comment by awaelchli** [comment]
Correct me if I'm wrong. This does solve #977. The discussion there is not over. As far as I can see this PR does not address the issue we have with restoring the hparams. This pr also adds other changes unrelated to hparams, why?

File: pytorch_lightning/core/lightning.py [hunk 8]
```diff
@@ -68,6 +68,21 @@ def __init__(self, *args, **kwargs):
...
+        if not hasattr(self, '_hparams'):
+            self._hparams = Namespace()
+        if isinstance(self._hparams, Namespace):
+            hparams = self._hparams
+        else:
+            assert isinstance(self._hparams, dict)
+            hparams = Namespace(**self._hparams)
+        return hparams
+
+    def set_hparams(self, params: Union[Dict[str, Any], Namespace]):
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py]: setters in python should be done with `@hparams.setter` decorator, no?

Comment (Borda) [reply pytorch_lightning/core/lightning.py]: can you in decorator define types?

**Comment by Borda** [comment]
> Correct me if I'm wrong. This does NOT solve #977. The discussion there is not over. As far as I can see this PR does not address the issue we have with restoring the hparams. This pr also adds other changes unrelated to hparams, why?

yeah, sorry, made is WIP, I have added somment to the original thread, could you check it pls...

File: pytorch_lightning/core/lightning.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py]:
```
@hparams.setter
def hparams(self, params: Union[Dict[str, Any], Namespace]):
    self._hparams = params
```
This should work. However, it does not enforce the types. The user can still pass in whatever it wants.

**Comment by awaelchli** [comment]
@Borda I see a small edge case (maybe). 
```
def __init__(params):
    self.myparams = params  # the user does not set self.params
    # later in code ...
    self.myparams['batch_size']  # the user works with a dict. Not allowed!
```

user wants to load the checkpoint now:
`model = MyLightningModule.load_from_checkpoint(...)  # this will pass in an empty Namespace to the module, because the checkpoint contains empty hparams Namespace
`
The model will not work. I guess this problem already existed before, but I think there was a warning when hparams was not set.

**Comment by williamFalcon** [comment]
great job guys! let's wrap this up in the next few hours so we can do the release. this is the last PR holding this up

**Comment by Borda** [comment]
@awaelchli I see a solution to have `self._hparams` as `dict` always and property `self.hparams` as Namespace so it does not break anything and if you need you may do
```
self.hparams.batch_size
self._hparams['batch_size']
```

**Comment by awaelchli** [comment]
Yes, that's one way. It would also mean from now on self.hparams is always going to be a Namespace, and it would break exising code bases where they had a dict or something else.

**Comment by Borda** [comment]
`E       urllib.error.HTTPError: HTTP Error 403: Forbidden` failing downalod dataset?

**Comment by awaelchli** [comment]
Had it earlier too. I added @williamFalcon 's links in PR #986 . That should solve the failing server problem when downloading mnist. hmm, I thought it was cached?

**Comment by Borda** [comment]
Just Linux machine was cached lok

**Comment by Borda** [comment]
nop,`torchvision` server with MNIST is down...

**Comment by williamFalcon** [comment]
@Borda 
This would break:
https://github.com/PyTorchLightning/pytorch-lightning/blob/f6934e5f14589ef3d35cfdae7e3c74eb3020637b/pytorch_lightning/core/lightning.py#L1198-L1229

I guess we don't have a test that saves and loads models with a dict?

**Comment by Borda** [comment]
@williamFalcon @awaelchli in checkpoint the `hparams` are always as `dict` or am I missing something?
https://github.com/PyTorchLightning/pytorch-lightning/blob/f6934e5f14589ef3d35cfdae7e3c74eb3020637b/pytorch_lightning/trainer/training_io.py#L240-L241

**Comment by awaelchli** [comment]
Yes there should probably a save and load test with dict. 
However, the user can always break the system, for example when accessing the arg before setting sefl.hparams = arg it could lead to typeerror.

**Comment by Borda** [comment]
FYI https://github.com/pytorch/vision/issues/1938

**Comment by williamFalcon** [comment]
this code forces a load as namespace. if the user model defined a dictionary then it would break no? because the user would access hparams[“arg”] but since we always force a namespace that would crash?

**Comment by awaelchli** [comment]
@williamFalcon if you want that behaviour I see no way around pickeling the whole hparams object or  saving the type with it as I suggested in #943 . Otherwise at load time we have no way of telling the type the user chose.

File: pytorch_lightning/trainer/training_io.py [hunk 9]
```diff
@@ -238,7 +238,9 @@ def dump_checkpoint(self):
         checkpoint['state_dict'] = model.state_dict()
 
         if hasattr(model, "hparams"):
-            checkpoint['hparams'] = vars(model.hparams)
+            is_namespace = isinstance(model.hparams, Namespace)
+            checkpoint['hparams'] = vars(model.hparams) if is_namespace else model.hparams
+            checkpont['hparams_type'] = 'namespaces' if is_namespace else 'dict'
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_io.py]: @awaelchli @Borda fixed it... now need to fix load

Comment (awaelchli) [reply pytorch_lightning/trainer/training_io.py]: how does that address your concern about hparams[".." ] access? load_from_checkpoint will pass a Namespace to Lightningmodule.

Comment (awaelchli) [reply pytorch_lightning/trainer/training_io.py]: ok nvm :)

Comment (Borda) [reply pytorch_lightning/trainer/training_io.py]: well I would rather stay with the fixed interface - `hparams` as Namespace property and setter

File: tests/models/base.py [hunk 10]
```diff
@@ -36,6 +36,28 @@ def __init__(self, root, train=True, transform=None, target_transform=None,
...
+        x, y = batch
+        y_hat = self.forward(x)
+        return {'loss': F.cross_entropy(y_hat, y)}
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(self.parameters(), lr=0.02)
+
+    def train_dataloader(self):
+        return DataLoader(MNIST(os.getcwd(), train=True, download=True,
+                                transform=transforms.ToTensor()), batch_size=32)
```
Comment (Borda) [inline tests/models/base.py]:
```suggestion
        return DataLoader(TestingMNIST(os.getcwd(), train=True, download=True,
                                       transform=transforms.ToTensor()), batch_size=32)
```

**Pull request merged**